### ISSUE TITLE
Add support for ARIA Live Region screen reader

### ISSUE THREAD

**Issue by esethna** [mattermost/mattermost#9729]
If you're interested please comment here and come [join our "Contributors" community channel](https://pre-release.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://pre-release.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-12764)

Original report: https://github.com/mattermost/mattermost-server/issues/9661

ARIA Live Region is not working properly - example page: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions.

**Comment by rexredinger** [comment]
I'll grab this one.

**Comment by hanzei** [comment]
Thanks @rexredinger

**Comment by rexredinger** [comment]
I'm sure this has never been said before, but accessibility is hard. 😸 
For starters, `FireFox + OSX VoiceOver + aria-live = Does Not Work` so I switched to ChromeVox. 

**Context:**
Developing against the webapp,
OSX (10.14.5),
Chrome Version 75.0.3770.90 (Official Build) (64-bit)
Using ChromeVox

I tried applying `role="log"` along with `aria-live="assertive"` to the `post-list` div.  It technically works!  But it also reads a bunch of things that it shouldn't, and in general highlights that a lot of other parts of the UI are not assistive-tech-friendly.

I am compiling a large list of changes to be made to establish a baseline for decent screen reader compat.  Expect that in the next day or so.

I am also still working on a PR that tries to make the minimal amount of changes needed to get a less annoying live zone for new messages.

**Comment by rexredinger** [comment]
New wrinkles.

Introducing a live zone means also having some clue about what has and hasn't been read. Why?

Because every update to the chat log, as it is implemented, represents a change of the entire log.  This means the whole thing gets reread after every new post.  To avoid rereading everything, we could push copies of new incoming posts to a live-region that is pushed off the page (it is standard practice to "hide" elements you want ARIA-accessible by just shoving them out of view or making them very small).  But this requires us to somehow know what has and hasn't been read by the screen reader and when to clear the live region (not really possible).

It may be simpler to have hidden live region that updates to alert the reader simply that new messages have been added and combine that with a ARIA landmarks that makes it easy for the operator of the reader to jump to the last message and navigate the chat log.

Of further note,  Slack only uses one live-region, `#aria_live_announcer`.  This is used to assertively announce what channel a user has jumped to.

PS: Still working on a triaged ARIA fixes/updates todo list.

**Comment by rexredinger** [comment]
**General**

- There are a lot of `divs` and nested `spans` that make it hard to navigate with a screen reader.  Flattening the DOM, where possible would be really helpful.
- More ubiquitous and semantic `aria-label` usage could go a long way toward making MM client screan readable.
- For buttons and links, try to use descriptive imperative language, describing what the button or link does.
- Most models, banners, and menus have an `x` button to dismiss.  These all need `aria-label="Dismiss"` or something similar set.  Right now, the screen reader either calls them "button" or "x", without any context.
- Slack has a lot of nice patterns established for supporting ARIA in their DOM.  It helps to looks at their web client when deciding how to approach things.
- This blog is also a great resource https://inclusive-components.design/menus-menu-buttons/ 
- For places we want the screen reader to focus in on first, it helps to set `tabindex='-1'`.
- Links should make request, or load new pages, not be responsible for transient page changes, like closing modals or expanding menus (use buttons for that).
- Even without any changes, Chrome Vox does pretty OK at navigation and giving mostly meaningful feedback.  However, Apple VoiceOver struggles.

_I would love some feedback on how something like JAWS on Windows interacts with MatterMost._

**Channel Header**

- The header buttons are all actually complex unlabeled `div`s. Consider marking the `flex-parent` `div` with `role='tablist'` and the `flex-child` elements with `aria-label`s and `role='tab'`

**Menus**

- Consider re-implementing menu container `div`s as more semantic `nav` elements
- Apply `role='presentation'` to  `div`s within the nav `div`, if they only exist for structural or representational purposes.  This will help to reduce navigation complexity.
- Menu-dividers as list items in a menu present a speed bump for screen readers.  Address with `aria-hidden='true'` property.
- Add appropriate `aria-labels` to all menu links/buttons.
- Set `aria-expandable='true'` for menus that are shown and `aria-expandable='false'` for those that are not.  This is already implemented dot-menu and team-members dropdown.

**LSB**

- Dynamically rewrite `aria-labels` for channels to note unread messages exist in that channel
- Minimize `div`s where a `span`, `p`, or `React.fragment` will do. `#headerUsername` is a good example.  Non-semantic div usage makes it harder for screen readers to prioritize focus on what matters most :wink:
- Refactor "Unreads", "Private Channels, "Public Channels", and "Direct Messages to be submenus.  Not as important but could really help with speeding up navigation
- Status menu could be labeled and refactored to have a top-level `nav` element.
- The Plus buttons need more descriptive labels.  The screen reader currently calls it "Plus", rather than something like "Add new public channel"

**RSB**

- I noticed (looking at the flag icon in the help text) that `aria-label` does not work to identify svg icons.  Chrome Vox just skips it entirely.  The label needs to be on the `span` that contains the svg.
- Otherwise RSB is looking pretty solid.

**Posting Box**

- The `#create_post` element is appropriately labeled with a form role, but needs an `aria-label`
- The hidden text area, `#post_textbox-reference` needs an `aria-hidden='true'`
- It might honestly be OK to 
- The `.post-body__actions` does not play well with the reader.  It would be great if these buttons could be refactored out of the nested `span`/`div` structure they are in now, and into some simpler, labeled buttons.

**Chat Log**

- I'm still trying to determine if there is a better way to present messages to make them more cleanly readable by a reader.  It seems even Slack hasn't solved this.

**Modals**

- All the modals would benefit from similar changes described above.  Most of it really comes back to adding aria-labels to things.

**Admin Console**

- From what I can see, a lot of the Admin console is also fairly navigable.
- Lots of form elements could still use labels.
- It would still be good to list the sidebar contained in a `nav` element


I think some of the more common controls and low-hanging fruit in this list should take priority.  Anyway, that's what I have for now.  I'll reiterate that it would be nice if someone with access to JAWS could share their experience interacting with the web client.

**Comment by hanzei** [comment]
@devinbinnie Can you please provided feedback on the above?

@rexredinger Sorry for the delay

**Comment by devinbinnie** [comment]
@rexredinger I've been using JAWS to implement a lot of accessibility changes since the beginning of July when I joined Mattermost, and the experience has improved a lot since then. @deanwhillier has implemented an Accessibility Navigation Controller (https://github.com/mattermost/mattermost-webapp/pull/3047) that makes navigation a lot easier, and myself and @asaadmahmood have added a lot of `aria` tags and changes to the DOM to address a lot of issues.

I've found JAWS to be one of the more difficult screen readers to use, partially due to its invasiveness into the operating system and its virtual cursor which causes navigation issues. However, most of our improvements have been tested using JAWS so it should be the most optimal screen reader to use currently.

Let me know if this is helpful, if you have any more specific questions I'd be happy to answer them :)

**Comment by hanzei** [comment]
@rexredinger Was the provided feedback help full?

**Comment by esethna** [comment]
@rexredinger thanks for your valuable insight here! Apologies for the delayed response, this ticket is relatively outdated as we made some major accessibility improvements (as Devin mentioned) in v5.14.

Would you be interested in syncing up with me and our devs so we can work more closely on how to make Mattermost even more accessible? Our goal is WCAG Level AA, and we'd love to work with individuals who have experience in the area to help get us there. You can mention me (@eric.sethna) in our ~Accessibility channel on our community server if you're interested. https://community-release.mattermost.com/core/channels/accessibility

Looking forward to learning more from you!

**Comment by KuSh** [comment]
Possible hints : https://a11y-guidelines.orange.com/en/articles/chatbot/

It seems the actual implementation with a dedicated sr-only aria-live element that contains the latest message only might not work in the event of a rapid influx of messages.

**Comment by Masarczyk-Wienfluss** [comment]
Greetings in 2024 and apologies if this issue is not the right place to start the conversation.

Is the Accessibility channel still the main place to discuss accessibility of Mattermost? Having revisited it after some time I see that there is very little activity on it and I would be interested to find out where Mattermost currently stands in regards to the accessibility goals.

I am not a dev myself but an accessibility tester and screen reader user and we employ Mattermost in our company so I would be happy to help out with any feedback I can.

**Comment by lieut-data** [comment]
Thanks for reporting this issue and for the community discussion around it. After multiple years of inactivity, we're closing this to help keep our issue tracker focused on active work. If this is still relevant, please feel free to open a new issue with updated context and we'll be happy to take another look.
